source/source.go

```diff
@@ -0,0 +1,227 @@
+	Next(context.Context) (sdk.Record, error)
+	Stop() error
+}
+
+// Source is an Amazon Redshift source plugin.
+type Source struct {
+	sdk.UnimplementedSource
+
+	config           config.Source
+	db               *sqlx.DB
+	lastProcessedVal any
+	iterator         Iterator
+}
+
+// NewSource initialises a new source.
+func NewSource() sdk.Source {
+	return sdk.SourceWithMiddleware(&Source{}, sdk.DefaultSourceMiddleware()...)
+}
+
+// Parameters returns a map of named Parameters that describe how to configure the Source.
+func (s *Source) Parameters() map[string]sdk.Parameter {
+	return map[string]sdk.Parameter{
+		config.DSN: {
+			Default:     "",
+			Required:    true,
+			Description: "data source name to connect to the Amazon Redshift.",
+		},
+		config.Table: {
+			Default:     "",
+			Required:    true,
+			Description: "Name of the table that the connector should read.",
+		},
+		config.OrderingColumn: {
+			Default:  "",
+			Required: true,
+			Description: "Column name that the connector will use for ordering rows. Column must contain unique " +
+				"values and suitable for sorting, otherwise the snapshot won't work correctly.",
+		},
+		config.CopyExistingData: {
+			Default:  "true",
+			Required: false,
+			Description: "The field determines whether the connector will move already existing data " +
+				"or only those that will be available after it starts.",
+		},
+		config.KeyColumns: {
+			Default:     "",
+			Required:    false,
+			Description: "Comma-separated list of column names to build the sdk.Record.Key.",
+		},
+		config.BatchSize: {
+			Default:     "1000",
+			Required:    false,
+			Description: "Size of rows batch. Min is 1 and max is 100000. The default is 1000.",
+		},
+	}
+}
+
+// Configure parses and stores configurations,
+// returns an error in case of invalid configuration.
+func (s *Source) Configure(ctx context.Context, cfgRaw map[string]string) error {
+	sdk.Logger(ctx).Info().Msg("Configuring Amazon Redshift Source...")
+
+	cfg, err := config.ParseSource(cfgRaw)
+	if err != nil {
+		return fmt.Errorf("parse source config: %w", err)
+	}
+
+	s.config = cfg
+
+	return nil
+}
+
+// Open prepare the plugin to start sending records from the given position.
+func (s *Source) Open(ctx context.Context, position sdk.Position) error {
+	sdk.Logger(ctx).Info().Msg("Opening an Amazon Redshift Source...")
+
+	var err error
+
+	s.db, err = sqlx.Open(driverName, s.config.DSN)
+	if err != nil {
+		return fmt.Errorf("open db connection: %w", err)
+	}
+
+	err = s.db.Ping()
+	if err != nil {
+		return fmt.Errorf("ping: %w", err)
+	}
+
+	// if position is nil will use it, if it's not and copyExistingData is false -
+	// populate it with the value of the last row orderingColumn column
+	if position != nil {
+		if err = json.Unmarshal(position, &s.lastProcessedVal); err != nil {
+			return fmt.Errorf("unmarshal sdk.Position into Position: %w", err)
+		}
+	} else if !s.config.CopyExistingData {
```

Comment:
as I understand if we don't need copy existing data we get last processed value by ordering column for reason that all data after that will be for cdc mode?

Reply:
I've changed this logic, so the comment is not relevant.